scripts/import_dvv.sql

```diff
@@ -21,21 +21,134 @@ insert into jkr_osoite.katu (katunimi_fi, katunimi_sv, kunta_koodi)
+        join jkr.rakennuksen_omistajat ro on r.id = ro.rakennus_id
+        join jkr.osapuoli op on ro.osapuoli_id = op.id
+        where r.prt = omistaja.rakennustunnus and op.nimi = omistaja."omistajan nimi"
+        ) -- Only add those names each building does not have listed as owners yet.
+          -- Note that this may introduce multiple owners with the same name for each building
+          -- if there are multiple such rows in the same file. They will still have different
+          -- addresses, though.
+;
+
+-- Insert owners to jkr.rakennuksen_omistajat
+-- Step 1: Find all buildings owned by each owner, matching by henkilötunnus
+insert into jkr.rakennuksen_omistajat (rakennus_id, osapuoli_id)
+select
+    (select id from jkr.rakennus where omistaja.rakennustunnus = rakennus.prt) as rakennus_id,
+    (select id from jkr.osapuoli where omistaja."henkilötunnus" = osapuoli.henkilotunnus) as osapuoli_id
+from jkr_dvv.omistaja
+where
+    omistaja."henkilötunnus" is not null and
+    exists (select 1 from jkr.rakennus where omistaja.rakennustunnus = rakennus.prt) -- not all buildings are listed
+on conflict do nothing; -- DVV has registered some owners twice on different dates
+
+-- Step 2: Find all buildings owned by each owner, matching by y-tunnus
+insert into jkr.rakennuksen_omistajat (rakennus_id, osapuoli_id)
+select
+    (select id from jkr.rakennus where omistaja.rakennustunnus = rakennus.prt) as rakennus_id,
+    (select id from jkr.osapuoli where omistaja."y_tunnus" = osapuoli.ytunnus) as osapuoli_id
+from jkr_dvv.omistaja
+where
+    omistaja."y_tunnus" is not null and
+    exists (select 1 from jkr.rakennus where omistaja.rakennustunnus = rakennus.prt) -- not all buildings are listed
+on conflict do nothing; -- DVV has registered some owners twice on different dates
+
+-- Step 3: Find all buildings owned by missing henkilötunnus/y-tunnus by name and address
+insert into jkr.rakennuksen_omistajat (rakennus_id, osapuoli_id)
+select
+    (select id from jkr.rakennus where omistaja.rakennustunnus = rakennus.prt) as rakennus_id,
+    (select id from jkr.osapuoli where
+        -- all fields must be equal or null to match.
+        -- some rows are exact duplicates. they should not be present in jkr.osapuoli.
+        omistaja."omistajan nimi" is not distinct from osapuoli.nimi and
+        omistaja."omistajan postiosoite" is not distinct from osapuoli.katuosoite and
+        omistaja."postiosoitteen postitoimipaikka" is not distinct from osapuoli.postitoimipaikka and
+        omistaja."postios posti_numero" is not distinct from osapuoli.postinumero and
+        omistaja."rakennustunnus" = osapuoli.rakennustunnus
+    ) as osapuoli_id
+from jkr_dvv.omistaja
+where
+    omistaja."henkilötunnus" is null and
+    omistaja."y_tunnus" is null and
+    exists (
+        select 1 from jkr.rakennus where omistaja.rakennustunnus = rakennus.prt) and -- not all buildings might be listed
+    not exists (
+        select 1 from jkr.rakennus r
```

Comment:
I think this last one is redundant condition since (rakennus_id, osapuoli_id) is unique on rakennuksen_omistajat  and you have the `on conflict do nothing` in place.

Reply:
Ah, didn't notice the (rakennus_id, osapuoli_id) primary key, good point.

Reply:
Having tested it without, this condition was needed after all. This is because we want the script to be repeatable and idempotent on the same data.

On the first run, this condition doesn't do anything, because it only filters those objects that have just been imported, and they contain the rakennustunnus data in the database.

On the second run, however, the rakennustunnus column is empty, because all of the buildings and owners have already been matched on the first run.

When the rakennustunnus column is empty, `(select id from jkr.osapuoli ...)` returns null for all ids that it is trying to insert, so actually we have incomplete data that we are trying to insert. The `osapuoli_id` not null constraint triggers right there. So we must filter out all the osapuolet that have rakennustunnus `null`.

Still, I will simplify this condition.

Reply:
You are right. Other maybe cleaner but maybe not faster way to do this is to select all rakennus_id, osapuoli_id pairs in subquery and then filter nulls out of it in outer query. But this is good enough also.